include/vikunja/mem/iterator/TransformIterator.hpp

```diff
@@ -20,212 +22,251 @@
 {
-    namespace mem
+    /**
+     * @brief A transform iterator, returning a value given initially at any index. As such it has no bounds,
+     * other than the bounds of the index type used.
+     * @tparam DataType The type of the data
+     * @tparam TInputIterator The type of the underlying input iterator
+     * @tparam TTransform The Functor transforming the input values
+     */
+    template<typename AccObj, typename DataType, typename TInputIterator, typename TTransform>
+    class TransformIterator
     {
-        namespace iterator
-        {
-            /**
-             * @brief A transform iterator, returning a value given initially at any index. As such it has no bounds,
-             * other than the bounds of the index type used.
-             * @tparam DataType The type of the data
-             * @tparam TInputIterator The type of the underlying input iterator
-             * @tparam TTransform The Functor transforming the input values
-             */
-            template<typename DataType, typename TInputIterator, typename TTransform>
-            class TransformIterator
-            {
-            public:
-                // Need all 5 of these types for iterator_traits
-                using difference_type = IdxType;
-                using value_type = DataType;
-                using pointer = DataType*;
-                using reference = DataType&;
-                using iterator_category = std::random_access_iterator_tag;
-
-                /**
-                 * @brief Constructor for the TransformIterator
-                 * @param it The underlying input iterator that this iterator will walk on and transform
-                 * @param func The function applied to the input iterator's values
-                 */
-                constexpr TransformIterator(const TInputIterator& it, const TTransform func) : it(it), func(func)
-                {
-                }
-
-                /**
-                 * @brief Dereference operator to receive the stored value
-                 */
-                NODISCARD constexpr ALPAKA_FN_INLINE const DataType& operator*() const
-                {
-                    return func(*it);
-                }
-
-                /**
-                 * @brief Index operator to get stored value at some given offset from this iterator
-                 */
-                NODISCARD constexpr ALPAKA_FN_INLINE const DataType& operator[](IdxType idx) const
-                {
-                    return func(v[idx]);
-                }
+    public:
+        using IdxType = int64_t;
+
+        // Need all 5 of these types for iterator_traits
+        using difference_type = IdxType;
+        using value_type = DataType;
+        using pointer = DataType*;
+        using reference = DataType&;
+        using iterator_category = std::random_access_iterator_tag;
+
+        using TInputIteratorDeref = decltype(*std::declval<TInputIterator>());
+        using AccObjPtr = AccObj*;
```

Comment:
I recognized, there is a big problem. At the moment, I don't know a way to determine the Acc type one the host side, before the algorithm is running. The concrete Acc type is injected by alpaka. Therefore, the type is first available, if the functor is called. I need to discuss the problem with the alpaka maintainer. Therefore, I suggest to implement the transform iterator without support for the acc object at the moment.